rasa/model.py

```diff
@@ -363,25 +353,40 @@ def should_retrain(new_fingerprint: Fingerprint, old_model: Text, train_path: Te
         to be retrained or not.
 
     """
-    retrain_nlu = retrain_core = True
+    retrain_nlu = retrain_core = replace_templates = True
 
     if old_model is None or not os.path.exists(old_model):
-        return retrain_core, retrain_nlu
+        return retrain_core, retrain_nlu, replace_templates
 
     with unpack_model(old_model) as unpacked:
         last_fingerprint = fingerprint_from_path(unpacked)
-
         old_core, old_nlu = get_model_subdirectories(unpacked)
 
-        if not core_fingerprint_changed(last_fingerprint, new_fingerprint):
+        retrain_core = section_fingerprint_changed(
+            last_fingerprint, new_fingerprint, SECTION_CORE
+        )
+        replace_templates = section_fingerprint_changed(
+            last_fingerprint, new_fingerprint, SECTION_TEMPLATES
+        )
+        retrain_nlu = section_fingerprint_changed(
+            last_fingerprint, new_fingerprint, SECTION_NLU
+        )
+
+        # If merging directories fails, force retrain.
+        if not retrain_core:
             target_path = os.path.join(train_path, "core")
             retrain_core = not merge_model(old_core, target_path)
-
-        if not nlu_fingerprint_changed(last_fingerprint, new_fingerprint):
+        else:
```

Comment:
I don't think the logic is correct here, as this else will be run if core needs to be retrained.

Reply:
I'm not sure I understood your concern here. Perhaps you can mention a specific combination of values for `retrain_core` and `retrain_nlg` that you think this will fail for?

Reply:
I think Tom means: why do we need to merge `core` in case `retrain_core=True`?
In general I don't like that `should_retrain` already does some modifications. I think that's unexpected behavior when I call a function `should_xy_be_done`. And I also think that's part of the reason why this seems so complicated (to me 😄 ).

Reply:
The docstring also indicates a totally different behavior of the function

Reply:
So the objectives are as follows:
1. `should_retrain.nlg` should return the value specific to whether *only* the NLG section has changed regardless of the value of  `should_retrain.core`. This is to ensure the returned values make logical sense.
2. Given that its value should be independent of `should_retrain.core`, we also want to check whether it's possible to merge the directories, because in the following case when Core needn't be retrained and merging directories is failing.

Reply:
But have updated the condition now to make it cleaner and less confusing.